testing/integration/upgrade_test.go

```diff
@@ -266,13 +266,71 @@ func TestStandaloneUpgradeWithGPGFallback(t *testing.T) {
 
 	_, defaultPGP := release.PGP()
 	firstSeven := string(defaultPGP[:7])
-	customPGP := strings.Replace(
+	newPgp := strings.Replace(
 		string(defaultPGP),
 		firstSeven,
 		"abcDEFg",
 		1,
 	)
 
+	customPGP := CustomPGP{
+		PGP: newPgp,
+	}
+
+	testStandaloneUpgrade(ctx, t, agentFixture, fromVersion, toVersion, "", false, false, true, customPGP)
+}
+
+func TestStandaloneUpgradeWithGPGFallbackOneRemoteFailing(t *testing.T) {
```

Comment:
This test is good for testing the standalone upgrade path. But it would be good to have a unit test that calls `verifier, err := newVerifier(...)` followed by calling `verifier.Verify(...)` with no `pgpBytes` being passed to `verifier.Verify(...)`, because this is what's happening in the Fleet-managed upgrade path - details below. The unit test could use the [testing logger](https://github.com/elastic/elastic-agent/blob/735a8fbc52c94b78f9b893b761e4a373df46c2e3/pkg/core/logger/testing.go#L15-L17) to assert that the new warning you added in this PR is actually emitted.

Looking at the Fleet-managed upgrade path, no `pgpBytes` are passed to the `coordinator.Upgrade(...)` call here: https://github.com/elastic/elastic-agent/blob/b19e95069e28d0389e2e66a8281e5df823566ecc/internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go#L78

Which means, no `pgpBytes` get passed to the `upgrader.Upgrade(...)` call here: https://github.com/elastic/elastic-agent/blob/b19e95069e28d0389e2e66a8281e5df823566ecc/internal/pkg/agent/application/coordinator/coordinator.go#L448

Which means, no `pgpBytes` are passed to the `u.downloadArtifact(...)` call here: https://github.com/elastic/elastic-agent/blob/b19e95069e28d0389e2e66a8281e5df823566ecc/internal/pkg/agent/application/upgrade/upgrade.go#L122

Which means, no `pgpBytes` are passed to the `verifier.Verify(...)` call here: https://github.com/elastic/elastic-agent/blob/b19e95069e28d0389e2e66a8281e5df823566ecc/internal/pkg/agent/application/upgrade/step_download.go#L84

However, right before that call, when `verifier` is [constructed](https://github.com/elastic/elastic-agent/blob/b19e95069e28d0389e2e66a8281e5df823566ecc/internal/pkg/agent/application/upgrade/step_download.go#L79), it ends up getting the embedded PGP bytes and passing them to each of three types of verifiers (fs, snapshot, and http): https://github.com/elastic/elastic-agent/blob/b19e95069e28d0389e2e66a8281e5df823566ecc/internal/pkg/agent/application/upgrade/step_download.go#L122-L145

So that's why everything works out, but it would be good to have a unit test that actually exercises this code path by calling `verifier, err := newVerifier(...)` followed by calling `verifier.Verify(...)` with no `pgpBytes` being passed to it. The unit test could use the [testing logger](https://github.com/elastic/elastic-agent/blob/735a8fbc52c94b78f9b893b761e4a373df46c2e3/pkg/core/logger/testing.go#L15-L17) to assert that the new warning you added in this PR is actually emitted.

Reply:
+1 to having a unit test, for an integration test I am more interested in a test that blocks access to the artifacts API entirely because that test will catch other attempts to reach out to the artifacts API unnecessarily, where as this one only catches attempts to get the PGP key.

Reply:
@ycombinator i see where you're getting but i think without passing pgpBytes we cannot verify much. we append remote pgp in step_download and we want to verify that we won't fail if this is unavailable. 
i'm passing fake uri as a pgpBytes here to simulate the situation with unreachable remote pgp, so i would need to pass something what definitely fails i order to check logger output.

i will add UT with this scenario still. 

@cmacknz i've seen the issue with integration test you're describing but as it was unassigned and unplanned i was treating these separately. that one will most likely require a bit more effort in a sense of client injections.